wgsl: support swizzle assignment

dneto0 opened the pull request:
Add language extension 'swizzle_assignment'

Example:

   var v:vec4u;
   v.yz = vec2(1,2);

Issue: #737

dneto0 committed:
wgsl: support swizzle assignment

Add language extension 'swizzle_assignment'

Example:

   var v:vec4u;
   v.yz = vec2(1,2);

Now includes fixes on the order of evaluation.

Issue: #737

dneto0 committed:
Fix the discussion in "Component Reference from Vector Memory View"

When using a single-letter swizzle name on a pointer or reference,
the result is a reference to a component. This can be used
anywhere a reference-to-component can be used!

dneto0 committed:
wgsl: support swizzle assignment

Add language extension 'swizzle_assignment'

Example:

   var v:vec4u;
   v.yz = vec2(1,2);

Now includes fixes on the order of evaluation.

Issue: #737

dneto0 committed:
Fix the discussion in "Component Reference from Vector Memory View"

When using a single-letter swizzle name on a pointer or reference,
the result is a reference to a component. This can be used
anywhere a reference-to-component can be used!

mehmetoguzderin commented on wgsl/syntax.bnf:
```diff
@@ -579,18 +579,18 @@ enable_extension_list :
 
 requires_directive :
-  'requires' software_extension_list ';'
+  'requires' language_extension_list ';'
```
@dneto0 Thank you very much for the draft. These diffs should not appear as they are already merged, a rebase/update against main would remove them. Thank you very much in advance for consideration.

dneto0 replied:
I was making this PR when I noticed that naming thing.  Then I make the renaming PR, and came back to this one.

I'm using the jj but am still getting used to it. I think I've properly rebased it now.

This is still a draft, so feel free to look around, but I'm not pushing this at committee yet. I haven't showed it to colleagues.  I think it's in pretty good shape, however.

mehmetoguzderin replied:
@dneto0 thank you! I like that swizzles are not a new/separate syntax rule but name (as they are valid names) in the grammar. Swizzles will make it easier to translate from other shader languages & vector math libraries, thanks a lot for the draft.

dj2 commented on wgsl/index.bs:
```diff
@@ -7096,16 +7782,40 @@ A phony-assignment is useful for:
+        3. Evaluate |e2|.
+
+        4. Compute |update| by performing the specified binary operation on |viewed_contents| and |e2|.
+            * For example, if the swizzle assignment is `v += w` then the specified operation is addition.
```
This isn't actually a swizzle assignment. Should this be `v.wx += w` to make clearer it's a swizzle assign?

dneto0 replied:
fixed in the latest draft.  Used `v.xy += w`

dneto0 reviewed (commented):
There is an interaction with 15.2.5 Function Scope Variable Value Analysis, where a swizzle view should sometimes be considered a [full reference](https://gpuweb.github.io/gpuweb/wgsl/#full-reference). It would be if the root variable is of vector type, because a swizzle assignment is deemed to be overwriting the entire vector.

sudonatalie commented on wgsl/index.bs:
```diff
@@ -3616,6 +3636,209 @@ and vice versa,
+**Input**:
+* [=Swizzle view=] (|p|, |IndexSeq|) of type swizzle&lt;|AS|,|S|,|AM|,|N|,|K|&gt;
+* Vector |old_contents| of type vec|N|&lt;|S|&gt;
+    * |old_contents| is intended to be the vector stored in the memory referenced by |p| at some point in the past.
```
I think we need to specify that `old_contents` is intended to be the vector stored in memory after `update` is evaluate (because the evaluation of `update` may have had side effects on the non-swizzled components of the vector)

sudonatalie commented on wgsl/index.bs:
In the case of a swizzle compound assignment the underlying memory is accessed thrice

dneto0 replied:
Should be resolved now.
see new HTML below.

dneto0 committed:
Fix note on num times compound assignment accesses memory

For reference, it's still twice.

For swizzle view, it's three times.

h/t @sudonatalie

dneto0 committed:
Fix note on num times compound assignment accesses memory

For reference, it's still twice.

For swizzle view, it's three times.

h/t @sudonatalie

alan-baker commented on wgsl/index.bs:
```diff
@@ -5899,15 +6141,16 @@ but a multi-letter swizzle expression always yields a value of [=vector=] type.
 
 Note: In the table above, [=reference types=] are implicitly handled via the [=load rule=].
 
-#### Component Reference from Vector Memory View #### {#component-reference-from-vector-memory-view}
```
Vector Multiple Component Selection reads a little oddly now:
> A multi-letter [swizzle](https://github.com/gpuweb/gpuweb/pull/5268/changes#swizzle) cannot appear on the left-hand side of an [assignment](https://github.com/gpuweb/gpuweb/pull/5268/changes#statement-assignment): The left-hand side of an assignment must be of [reference type](https://github.com/gpuweb/gpuweb/pull/5268/changes#reference-type), but a multi-letter swizzle expression always yields a value of [vector](https://github.com/gpuweb/gpuweb/pull/5268/changes#vector) type.

alan-baker reviewed (changes requested):
I think you're missing an entry for the LHS expression uniformity rules for swizzle views.

dneto0 committed:
apply some feedback from @alan-baker

dneto0 committed:
apply some feedback from @alan-baker

dneto0 committed:
Clean up case breakdown in multi-component vector selection

dneto0 committed:
Clean up case breakdown in multi-component vector selection

dneto0 committed:
In the multi-component vector decomposition section,
split the tables between the case where the inner expression
is a vector vs. a memory view of a vector.

dneto0 committed:
In the multi-component vector decomposition section,
split the tables between the case where the inner expression
is a vector vs. a memory view of a vector.

dneto0 committed:
Refine the statement about not allowing swizzles on the LHS of an
assignment.

dneto0 committed:
Refine the statement about not allowing swizzles on the LHS of an
assignment.

dneto0 committed:
Describe originating variable of reference-from-swizzle view via single
letter.

dneto0 committed:
Swizzle views are not concrete

dneto0 committed:
Swizzle views are not concrete

dneto0 committed:
Describe originating variable of reference-from-swizzle view via single
letter.

dneto0 committed:
Table for swizzle view from memory view: add condition on swizzle_assignment

dneto0 committed:
Table for swizzle view from memory view: add condition on swizzle_assignment

dneto0 committed:
Define full/partial swizzle view and update uniformity statement rules

dneto0 committed:
Define full/partial swizzle view and update uniformity statement rules

dneto0 committed:
conversion rank: move swizzle view load rule to be beside the load rule

dneto0 committed:
conversion rank: move swizzle view load rule to be beside the load rule

dneto0 committed:
Tighten wording in component reference from Vector Memory View or Swizzle View

The result is always a reference, not just a memory view.

dneto0 committed:
Tighten wording in component reference from Vector Memory View or Swizzle View

The result is always a reference, not just a memory view.

dneto0 committed:
better linking

dneto0 committed:
better linking

dneto0 committed:
Reads and Writes happen during swizzle assignment

dneto0 committed:
Reads and Writes happen during swizzle assignment

dneto0 committed:
Split load-rule/swizzle-view-load-rule lines in uniformity-on-expressions table

  We need to splite MVE memory-view expression from SVE
  swizzle-view-expression

dneto0 committed:
Split load-rule/swizzle-view-load-rule lines in uniformity-on-expressions table

  We need to splite MVE memory-view expression from SVE
  swizzle-view-expression

dneto0 committed:
Split e.field and e.swiz rows in uniformity-for-expressions table

dneto0 committed:
Split e.field and e.swiz rows in uniformity-for-expressions table

dneto0 committed:
Fix unenclosed </div>

dneto0 committed:
Fix unenclosed </div>

alan-baker commented on wgsl/.python-version:
```diff
@@ -0,0 +1 @@
+3.12
```
Is this necessary?

dneto0 committed:
Add the swizzle case to LHSValue unfiormity rules

dneto0 committed:
Delete stray .python-version file

dneto0 committed:
Remove stray .python-version

dneto0 committed:
Add the swizzle case to LHSValue unfiormity rules

dneto0 committed:
Delete stray .python-version file

dneto0 committed:
Remove stray .python-version

dneto0 commented:
I'm happy with this draft.

My team has implemented this in Tint (behind a flag), and written CTS tests for it, *except* for the changes in uniformity analysis. We'll be doing that part soon.  Implementation point of contact is @sudonatalie

alan-baker commented:
@jimblandy Is mozilla ok with this feature?

jimblandy commented on tools/check-nfkc.py:
```diff
@@ -10,7 +10,7 @@
 
 files = subprocess.run(
-    ["git", "ls-files"], check=True, capture_output=True
+    ["git", "ls-files", "--no-deleted"], check=True, capture_output=True
```
Is this change intended to be part of this PR?

dneto0 replied:
yes.  I needed it locally at least temporarily.
It has been landed independently to main.  Upon rebase it will disappear.

jimblandy commented on wgsl/index.bs:
```diff
@@ -2570,16 +2590,17 @@ Note: When no conversion is performed, the conversion rank is zero.
-automatic conversion will only ever produce an [=f16=] if
-[=extension/f16|extension=] is enabled in the module.
+automatic conversion will only ever produce an [=f16=] value if
+the [=extension/f16|f16 extension=] is enabled in the module.
```
This change also seems irrelevant to the PR.

dneto0 replied:
yes.

It's a drive-by.  I've posted it independently as https://github.com/gpuweb/gpuweb/pull/6288
Once that lands, a rebase of this PR will make it disappear here.

jimblandy commented on wgsl/index.bs:
```diff
@@ -3637,6 +3658,212 @@ and vice versa,
+
+<blockquote algorithm="writing via a swizzle view">
+
+**Algorithm:** <dfn lt="swizzle view write">SwizzleViewWrite</dfn>(|p|, |IndexSeq|, |old_contents|, |update|)
```
Editorial: The `SwizzleViewRead` algorithm doesn't list its arguments above; should it? Or should we not list the arguments here?

Do these arguments match those listed in the `**Input**` section below? It seems like the `SwizzleView` is a single value, but the list here gives `p` and `IndexSeq` as distinct values.

jimblandy replied:
I see how the argument list corresponds with the `**Input**` section now. Might it be clearer to just have it take a single `sv` argument, and then use the "by convention" trick in the body?

dneto0 replied:
Hm.  SwizzleViewWrite is the exception. All the other "Algorithms" in the spec don't use a parameter list like this.  I'll delete it from SwizzleViewWrite.

jimblandy commented on wgsl/index.bs:
```diff
@@ -3637,6 +3658,212 @@ and vice versa,
+even if component values are unchanged by the process.
+
+**Input**:
+* [=Swizzle view=] (|p|, |IndexSeq|) of type swizzle&lt;|AS|,|S|,|AM|,|N|,|K|&gt;
```
Should `SwizzleViewWrite` assert that `AM` is `write` or `read_write`?

dneto0 replied:
It should assert AM is read_write.    We must be able to both read and write.

dneto0 replied:
The definition of teh swizzle view type has a clause:

> * (p,IndexSeq) supports a [swizzle view write](https://github.com/gpuweb/gpuweb/pull/5268#swizzle-view-write) operation that writes a vector of type vecK&lt;S&gt;, if AM is [read_write](https://github.com/gpuweb/gpuweb/pull/5268#access-read_write) or [write](https://github.com/gpuweb/gpuweb/pull/5268#access-write).

But it's best to reinforce it in the algorithm itself.

dneto0 replied:
So, interesting point.  The swizzle view type implies that `write` is an acceptable access mode.  But it's not usable, since we always have to be able to read. 

Fortunately variables that can be used with swizzle views are always read or read-write.  (All the non-texture address spaces).

So I'll remove the ability to use `write`.

dneto0 replied:
i made this change in commit "swizzle view access mode must be read or read_write"

jimblandy commented on wgsl/index.bs:
```diff
@@ -3637,6 +3658,212 @@ and vice versa,
+<div algorithm="swizzle view">
+A <dfn noexport>swizzle view</dfn> is a pair (|p|,|IndexSeq|) where:
+* |p| is a pointer to a vector of type vec|N|&lt;|S|&gt;
+* |IndexSeq| = |Idx|<sub>0</sub>...|Idx|<sub>|K|&minus;1</sub>, a sequence of indices where:
```
Editorial: I don't know if this is too W3C-ish, but rather than using the term "sequence", I think we should consider "list".

"Sequence" has [a definition](https://webidl.spec.whatwg.org/#idl-sequence) in IDL that might be fine, but I don't think is what we mean. This spec mostly uses "sequence" when talking about source text.

The [list](https://infra.spec.whatwg.org/#lists) concept from the Infra standard seems to be closer to what we want. The `SI(I)SI(J)SI(K)` concatenation syntax used to build IndexSeq values in the Swizzle View Expression is needlessly novel. It seems better to just use the `«SI(I), SI(J), SI(K)»` syntax from Infra. Infra lists have indexing syntax and plenty of useful verbs.

dneto0 replied:
TIL.  

I've made the change, in commit "Use W3C "list" notation instead of making up index sequence"

jimblandy commented on wgsl/index.bs:
```diff
@@ -6765,6 +7162,385 @@ The <dfn noexport>indirection</dfn> operator converts a pointer to its correspon
+</table>
+
+<table class='data'>
+  <caption>Forming a swizzle view from a swizzle view</caption>
```
Would it be helpful to have an assertion in the descriptions that `InputK` equals `IndexSeq`'s size?

dneto0 replied:
This is true by notation.

On each clause after the naming of `sv`, I've added "By definition, |IndexList| has |InputK| items.".

This is commit named  "Assert IndexList has InputK items."

jimblandy commented on wgsl/index.bs:
```diff
@@ -7162,6 +7939,87 @@ See [[#forming-references-and-pointers]] for other cases.
+    <td>By convention |sv| = (|p|, |IndexSeq|)
+
+        Constraints:
+        * The indices in |IndexSeq| [=shader-creation error|must=] be all distinct from each other.
```
Probably this comment should not result in a change, but I wanted to at least mention the idea:

Here and in the compound assignment rules, we require that `AM` must allow writes, and that the indices must be distinct from each other. But is there any reason to allow swizzle views with writable access modes to have non-distinct indices at all? Wouldn't it make more sense to check the indices for distinctness when forming views, and remove writability at that point?

The answer is: it's better to report the error when you try to assign to the swizzle than to have read-write swizzle views implicitly demoted to read-only swizzle views, and then wonder why you're trying to assign to a read-only swizzle view. Of course implementations don't have to work this way, but this way of specifying it hints at a reasonable way to implement it.

dneto0 replied:
I had not thought of it that way.

I don't like changing the type of an expression based on its enclosing use. That goes against the  design decision that WGSL infers types in a purely bottom-up way. (We go through a lot of effort to define abstract numeric types in such a way as to defer implicit conversions until cross-checking uses of expressions within a phrase; but that doesn't change the expressions' type, per se. I hope that makes sense).

Also it doesn't impact any use cases because there is no way to capture a swizzle view as a first class value.  That is, you can't make a let have a swizzle-view type. 

So I'm against thinking of the rules this way.

jimblandy commented on wgsl/index.bs:
```diff
@@ -6765,6 +7162,385 @@ The <dfn noexport>indirection</dfn> operator converts a pointer to its correspon
+* SI(`a`) &equals; 3
+
+<table class='data'>
+  <caption>Forming a swizzle view from a reference</caption>
```
Editorial: These tables are very long and redundant. Isn't there some more compact way to capture the patterns here that is still legible?

The right helper abstractions might make a shorter form of these tables *more* legible. It seems like something that deals with the separation of `xyzw` from `rgba`, the length constraints implied by the letters present, and so on could go a long way.

It does seem that the pointer, reference, and swizzle view cases must be separate tables.

(I didn't actually catch any mistakes in the tables.)

dneto0 replied:
I found several typo-level mistakes while editing for other items. :-)

I'm all ears on how to make these tables more legible *and* concise.  Everything I've thought of so far introduces too much abstraction and makes it hard to understand IMHO.

jimblandy commented on wgsl/index.bs:
```diff
@@ -7234,16 +8092,46 @@ A phony-assignment is useful for:
-* the reference expression |e1| is evaluated only once, and
-* the [=reference type=] for |e1| [=shader-creation error|must=] have a [=access/read_write=] [=access mode=].
+
+* If the expression |e1| is a [=reference type|reference=] expression:
```
It's inconsistent to spell out the evaluation steps when e1 is a swizzle view, but not when e1 is a reference. Rather than the special note about the Load rule, it seems like providing step-by-step rules for references as well would be clearer, and make it easier to compare the two cases.

This might also make the Note below about e1 being evaluated once but read twice unnecessary, since it'll be apparent in the algorithm.

dneto0 replied:
done in  commit "Elaborate steps in compound assignment via reference"

jimblandy commented on wgsl/index.bs:
```diff
@@ -3788,6 +4023,17 @@ References and pointers are distinguished by how they are used:
+    * In a function, when a [=swizzle view=] |sv| is used in a statement or an expression, where:
+    * |sv| is (|p|, |IndexSeq|) and of type swizzle&lt;|AS|,|S|,|AM|,|N|,|K|&gt;, and
+    * Access mode |AM| is [=access/read=] or [=access/read_write=], and
+    * The only potentially matching type rules require |sv| to have a value of type vec|K|&lt;|S|&gt;, then
```
I am a little surprised these three list items are not indented to fall underneath the "where:" clause. Don't they form the prerequisites for "That type rule requirement is considered to have been met"?

I guess the same comment would apply to the ordinary Load Rule text above.

dneto0 replied:
sure.

Mechanically doing this leaves a trailing "then" on the last of the sub-bullets. I have moved it to the start of the next clause.  "Then that type rule..."

this is done in commit "Indent as needed in Load Rule and Swizzle View Load Rule"

jimblandy commented on wgsl/index.bs:
```diff
@@ -3637,6 +3658,212 @@ and vice versa,
+      // Same as: v.zy = vec2(99,100);
+      v.yz.yx = vec2(99,100);
+
+      // Swizzles convert to a reference to a scalar component.
```
This is an odd way to put it. Referring to a single component of a swizzle yields a reference to the scalar component, because there is a special evaluation rule that says so, not because any conversion takes place.

dneto0 replied:
rewritten as

>  // Applying a single letter on a swizzle view yields a references to a scalar
  // component of the underlying vector.

done in commit "Rewrite comment about applying single letter to swizzle view"

jimblandy commented on wgsl/index.bs:
```diff
@@ -3637,6 +3658,212 @@ and vice versa,
+      // Same as: p.zy = vec2(99,100);
+      p.yz.yx = vec2(99,100);
+
+      // Swizzles convert to a pointer to a scalar component.
```
Similarly: this doesn't match the way the rest of the spec describes the meaning of that expression. It is more correct to say that swizzle views can be obtained from pointers, just as they can from references, and then referring to a single component proceeds as with any other swizzle view. Granted, this comment shouldn't be that verbose - but it shouldn't flat out say stuff that's wrong.

dneto0 replied:
It's fine to be verbose.

New text:

>   // A swizzle view can be constructed from a pointer.
  // Applying a single letter to a swizzle view yields a reference
  // to a scalar component of the underlying vector.


done in commit "Rewrite comment about applying single letter to swizzle view"

jimblandy commented on wgsl/index.bs:
```diff
@@ -3637,6 +3658,212 @@ and vice versa,
+    fn invalid_swizzle_cases() {
+      var v: vec2u;
+
+      // Invalid: Must not repeat the target component
```
This is phrased obliquely. It is certainly valid to repeat components in swizzles, just not to assign to such swizzles. Suggest:

```
// Invalid: cannot assign to swizzles that repeat components
```

dneto0 replied:
done in commit "Fix comment about assigning to swizzle with repeated components"

jimblandy commented on wgsl/index.bs:
```diff
@@ -3637,6 +3658,212 @@ and vice versa,
+      // Here v.xy is a swizzle value. There is no array-indexing expression on
+      // swizzle views, so the Swizzle Load Rule applies, yielding v.xy as a value
+      // of type vec2u. Then the indexed expression v.xy[0] results in a scalar value
+      // of type u32.  That cannot be the left hand side of an assignment, so
```
Suggest: "... as it is not a reference, so"

dneto0 replied:
reworded:

>   // of type vec2u. Then the indexed expression v.xy[0] results in a scalar value
  // of type u32. That cannot be the left hand side of an assignment, since it is
  // not a reference. Therefore, type checking fails.


done in commit "Elaborated comment in "invalid swizzle views" example"

jimblandy reviewed (commented):
I haven't finished reviewing everything, but here are the comments I have so far.

jimblandy commented on wgsl/index.bs:
```diff
@@ -3788,6 +4023,17 @@ References and pointers are distinguished by how they are used:
+    * In a function, when a [=swizzle view=] |sv| is used in a statement or an expression, where:
+    * |sv| is (|p|, |IndexSeq|) and of type swizzle&lt;|AS|,|S|,|AM|,|N|,|K|&gt;, and
+    * Access mode |AM| is [=access/read=] or [=access/read_write=], and
+    * The only potentially matching type rules require |sv| to have a value of type vec|K|&lt;|S|&gt;, then
```
I don't think we can say "the *only* potentially matching type rules require a type of vec*K*" here. The swizzle view load rule should apply to `e` in `e.x` or `sin(e)`, and the type rules for both of those permit many different vector types. Suggest:

> There exist potentially matching type rules that require `sv` to have a value of type `vecK<S>`

dneto0 replied:
I think that's not right.

For `e.x` case, the Swizzle View Load Rule should *not* apply.  Instead, apply a " Component Reference from Vector Memory View or Swizzle View" rule, with

     e: swizzle view
     e.x : reference to the vector component in memory.

For the `sin(e)` case, `sin` doesn't take a swizzle view argument, only a scalar or vector argument.  The vector case applies, and so the Swizzle View Load Rule applies:  perform a swizzle view to load the vector and form the new vector components via a shuffle, then pass that result to `sin`.

jimblandy replied:
You're right about `e.x`.

I think you're right about `sin(e)` as well, since the rule given at the beginning of "6.1.2 Conversion Rank" says that a rule's precondition is satisfied if the types match or if the actual type is automatically convertible to the required type. This is what determines the set of overloads. Since `swizzle<..., K>` only automatically converts to `vecK<...>`, I guess we can say "only" here.

jimblandy commented on wgsl/index.bs:
```diff
@@ -5850,111 +6096,201 @@ The result type depends on the number of letters provided. Assuming a `vec4<f32>
-A multi-letter [=swizzle=] cannot appear on the left-hand side of an [=statement/assignment=]:
+Note: If the [=language_extension/swizzle_assignment=] language extension is not supported,
+then a multi-letter [=swizzle=] cannot appear on the left-hand side of an [=statement/assignment=]:
 The left-hand side of an assignment must be of [=reference type=],
```
[Edit: I misread this.]<del>This claim is now out of date. The left-hand side of an assignment may be of either reference type or swizzle view type.</del>

jimblandy reviewed (commented):
GitHub makes it hard to comment on text that isn't affected by a PR, but there are a few places that I think need updates:

- In section 6, we have a note that says, "Reference types are not written in WGSL modules." Should that also mention Swizzle View types?

- The section on reference types mentions that they must not appear in WGSL source. Should the section on Swizzle View types do the same?

- Section "Use cases for references and pointers" says that "The left-hand side of the assignment statement must be of reference type". This is no longer true.

-   There are many places that mention memory views which should now also mention swizzle views, since swizzle views are not considered memory views:

    - The section "Forming reference and pointer values" says, "Given a memory view with a vector store type, appending a single-letter vector access phrase results in a reference to the named component of the vector."

    - The introduction to Section "Composite Value Decomposition Expressions" uses "memory view" in the introductory paragraph, and shortly thereafcter claims that "named component expression"s can only be applied to composite values or memory views to such values."

    It does seem to me that we have to keep memory views and swizzle views separate, though: at the very least, we need some term for what `&` can be applied to.

jimblandy commented on wgsl/index.bs:
```diff
@@ -3637,6 +3658,212 @@ and vice versa,
+Write components of an |update| vector into specified components of the vector in memory referenced by pointer |p|.
+
+Note: Writing via a swizzle view writes the entire vector in memory,
+even if component values are unchanged by the process.
```
This seems to be in tension with the text of §15.2.5, Function Scope Variable Value Analysis, which distinguishes between "full swizzle views" and "partial swizzle views", and treats assignments to the latter like partial assignments. If SwizzleViewWrite always writes the entire vector, then it seems the uniformity analysis should be able to treat any assignment to a swizzle view, full or partial, as a full assignment.

dneto0 replied:
You're right there is a tension.

Even with writing the whole vector, the uniformity analysis still profits from the full / partial swizzle views. The partial views are still partial because the non-updated components remain the same.  

We need this note because of the interaction with the memory model.  Two different invocations will race even if they write to different components of the vector.

e.g.  for workgroup variable `w`:

     thread 0:        w.xy = vec2u(0,1);
     thread 1:        w.zw = vec2u(2,3);

These threads race because  both writes will touch all 4 components.

I should expand the note with the above explanation.

dneto0 replied:
I've updated the text for this, and pushed the branch.

jimblandy replied:
Okay, so in other words, the spec was correct:

- Even though `SwizzleViewWrite` writes the entire vector,  its load-modify-store maneuver still blends the uniformities of the vector's modified and unmodified components. This means the uniformity analysis must treat a store to a partial swizzle view as a partial assignment.

- Usually memory models forbid stores that aren't apparent in the code. (For example, threads are not allowed to arbitrarily load from a global and then store the loaded value back into it, unless that's what the source code says to do, since other threads might be able to see such arbitrary stores.) However, swizzle assignment does store to components that are not named in the code, and data races need to be allowed to observe these stores.

Have I got that right?

dneto0 replied:
yes. that's exactly right.

In fact we already have a rule about the interaction with the memory model when writing to a vector component.  The implementation is allowed to touch the entire vector.

From [8.5.1.3 Component reference from vectory memory view or swizzle view](file:///Users/dneto/project/gpuweb/wgsl/index.html#component-reference-from-vector-memory-view)

> A [write access](https://github.com/gpuweb/gpuweb/pull/5268#write-access) to a component of a vector may access all of the [memory locations](https://github.com/gpuweb/gpuweb/pull/5268#memory-locations) associated with that vector.

Then there is a following note.

> Note: This means accesses to different components of a vector in memory by different invocations must be synchronized if at least one access is a [write access](https://github.com/gpuweb/gpuweb/pull/5268#write-access). See [§ 17.11 Synchronization Built-in Functions](https://github.com/gpuweb/gpuweb/pull/5268#sync-builtin-functions).

I'm happy to take feedback on how to combine, better call out, or otherwise enhance the language around this in the spec.

jimblandy commented on wgsl/index.bs:
```diff
@@ -7162,6 +7939,87 @@ See [[#forming-references-and-pointers]] for other cases.
       </xmp>
     </div>
 
+### Swizzle Assignment ### {#swizzle-assignment-section}
```
Since assignment is the whole point of the language extension, would it be appropriate to mention it here?

dneto0 replied:
I assume the second "it" is the language extension?  That is, add a sentence to say this section only applies when the language extension is supported.

dneto0 replied:
done, and pushed to the ranch.

dneto0 committed:
editorial: Link the "Swizzle Assignment" section to the language extension

dneto0 committed:
Explain the interaction of SwizzleViewWrite with uniformity and the memory model

dneto0 committed:
delete parameter list from "Algorithm: SwizzleViewWrite"

That makes it consistent with other "Algorithm" blocks.

dneto0 committed:
swizzle view access mode must be read or read_write

SwizzleViewWrite is only usable with read_write swizzle views

dneto0 committed:
Use W3C "list" notation instead of making up index sequence

Also, rename IndexSeq to IndexList

dneto0 committed:
Assert IndexList has InputK items.

This is in the "swizzle view from swizzle view" table.

dneto0 committed:
Elaborate steps in compound assignment via reference

This gives it structure parallel to compound assignment via swizzle
view, so they are easier to compare.
It removes a possibly-confusing reference to using the load rule.

dneto0 committed:
Indent as needed in Load Rule and Swizzle View Load Rule

dneto0 committed:
Rewrite comment about applying single letter to swizzle view

dneto0 committed:
Fix comment about assigning to swizzle with repeated components

dneto0 committed:
Elaborated comment in "invalid swizzle views" example

dneto0 committed:
Add note that swizzle view types are not spelled out in WGSL source

dneto0 committed:
Add swizzle view cases to "use cases for references and pointers"

dneto0 committed:
add missing parallel cases to memory-view, outside of the memory model section

dneto0 committed:
add missing parallel cases to memory-view:  memory model section

dneto0 committed:
In "use cases for references and pointers", further qualify
rules about simple assignments as being about *simple* assignment
(as opposed to swizzle assignment).

jimblandy commented on wgsl/index.bs:
```diff
@@ -3637,6 +3661,243 @@ and vice versa,
+
+**Procedure:**
+
+1. Read vector |u| from memory referenced by |p|.  |u| has type type vec|N|&lt;|S|&gt;.
```
This algorithm should probably have a "By convention IndexList equals ...", like SwizzleViewWrite, since it does refer to `Idx_i`.

jimblandy replied:
```suggestion
1. Read vector |u| from memory referenced by |p|.  |u| has type vec|N|&lt;|S|&gt;.
```

jimblandy commented on wgsl/index.bs:
```diff
@@ -3637,6 +3661,243 @@ and vice versa,
+      // Same as: v.zy = vec2(99,100);
+      v.yz.yx = vec2(99,100);
+
+      // Applying a single letter on a swizzle view yields a references
```
```suggestion
      // Applying a single letter on a swizzle view yields a reference
```

jimblandy commented on wgsl/index.bs:
```diff
@@ -3637,6 +3661,243 @@ and vice versa,
+  </xmp>
+</div>
+
+For the purposes of uniformity analysis,
+a [=swizzle view write=] may only affect the value uniformity of the updated components.
+This motivates the distinction between [=full swizzle views=] and [=partial swizzle views=].
+
+However, writing all components of the underlying vector is significant for the memory model.
+Two different invocations that concurrently update different components of the same vector in memory
+will induce a data race.
```
This is great. The example is very helpful.

jimblandy commented on wgsl/index.bs:
```diff
@@ -3637,6 +3661,243 @@ and vice versa,
+        and vec|K|&lt;|S|&gt; are valid [=vector=] types.<br>
+        ptr&lt;|AS|,vec|N|&lt;|S|&gt;,|AM|&gt; is a valid pointer type,
+        where |AM| is [=access/read=] or [=access/read_write=].<br>
+        If |AM| is [=access/read_write=] then additionally |K| &le; |N|.
```
I don't understand why this would be true. The type rules in "8.15. Swizzle View Expression" carry the access mode over directly from the base reference or pointer, so if `r` is `ref<function, vec2<f32>, read_write>`, then those rules say that `r.xxx` has type `swizzle<function, f32, read_write, 2, 3>`, in which `K` is greater than `N`.

dneto0 replied:
you're right. That should be deleted.  Thanks!

I was trying to constrain the bad cases for swizzle assignment as early as possible in the logical dependency tree.  But that's too far.  There are three places where the uniqueness constraint on indices is already stated, and in the input description for SwizzleViewWrite, it spells out the inference that |K| &leq; |N|.

jimblandy reviewed (commented):
Review still in progress, comments so far:

jimblandy commented on wgsl/index.bs:
```diff
@@ -5864,97 +6164,181 @@ but a multi-letter swizzle expression always yields a value of [=vector=] type.
-    In this case the [=Load Rule=] applies, loading the vector component from memory and yielding that component as the result.
+* any place where the reference to a vector component type can appear.
+    Sometimes the [=Load Rule=] will apply,
+    loading the vector component from memory and yielding that component as the result.
```
I don't understand this change. The original text is clearer and still seems to apply. Perhaps this is better?


```suggestion
* any place where a reference to a vector component can appear.
    Sometimes the [=Load Rule=] will apply,
    loading the vector component from memory and yielding that component as the result.
```

dneto0 replied:
This change is needed because you can form reference  `v.x` and then expand that out to a swizzle view via `v.x.xxxx`

But yes, your suggestion is good, and I'll take it.

dneto0 replied:
My wording tried to shift it away from *values* to types since the sentence is about type rules and type checking and inference.  But it reads clearer your way.

jimblandy committed:
Introduce a definition to shorten Swizzle View Expression tables.

Introduce the definition "The swizzle Z represents the K indices
«Idx_0, ..., Idx_(K-1)»", and then use that to capture repetitive
structure in the rules for swizzle view expressions.

jimblandy reviewed (commented):
I took a shot at simplifying the Swizzle View Expression rule tables. It's available here:

https://github.com/jimblandy/gpuweb/tree/swizzle-represents-indices

Specifically, this commit makes the change: https://github.com/jimblandy/gpuweb/commit/c43f2dc4ea7f3283d041a8f3c440f3089f5790ad

Previews here: https://github.com/gpuweb/gpuweb/pull/6323

dneto0 committed:
Add "By convention IndexList = ..." to SwizzleViewRead

Also fix typos

dneto0 committed:
In swizzle view construction constrain K in {2,3,4}

The important thing is to exclude K=1.

By the grammar rule for swizzle_name, the swizzle name
can have a single letter.  Remove all doubt about the possibility
for constructing an invalid swizzle view, by adding the precondition
on K.

dneto0 committed:
Fix wording for when reference-to-vector-component can appear

dneto0 committed:
swizzle-view with read_write access mode can have  K > N

jimblandy commented on wgsl/index.bs:
```diff
@@ -6778,6 +7228,112 @@ The <dfn noexport>indirection</dfn> operator converts a pointer to its correspon
 
+## Swizzle View Expression ## {#swizzle-view-expr}
+
+Swizzle views expressions are enabled by the [=language_extension/swizzle_assignment=] language extension.
```
```suggestion
Swizzle view expressions are enabled by the [=language_extension/swizzle_assignment=] language extension.
```

jimblandy commented on wgsl/index.bs:
```diff
@@ -3650,6 +3674,244 @@ and vice versa,
+Note: Reading via a swizzle view reads the entire vector from memory, even if some
+vector components do not appear in the result.
+
+**Input**: [=Swizzle view=] (|p|, |IndexList|)
```
It would be nice to spell out the type of this argument, as we do for `SwizzleViewWrite`. Then `N`, `S`, and `K` in the procedure would be clearer. Probably it would be good to spell out the output type, too.

jimblandy commented on wgsl/index.bs:
```diff
@@ -5863,12 +6158,18 @@ The result type depends on the number of letters provided. Assuming a `vec4<f32>
 
-The expressions in this section are all multi-letter [=swizzles=].
-Each forms a [=vector=] from the components of another vector.
+The expressions in this section apply a multi-letter [=swizzles=] to an inner expression:
```
```suggestion
The expressions in this section apply a multi-letter [=swizzle=] to an inner expression:
```

jimblandy commented on wgsl/index.bs:
```diff
@@ -7175,6 +7732,89 @@ See [[#forming-references-and-pointers]] for other cases.
+
+        Constraints:
+        * The indices in |IndexList| [=shader-creation error|must=] be all distinct from each other.
+        * Each index in |IndexList| [=shader-creation error|must=] be in [0,|N|-1].
```
This constraint is already enforced by the "Swizzle View Expression" rules. Do we need to repeat it here? Should we just assert it instead?

dneto0 replied:
I made this a note instead of a constraint.

jimblandy commented on wgsl/index.bs:
```diff
@@ -7247,16 +7887,59 @@ A phony-assignment is useful for:
+    * By convention swizzle view |e1| can be written (|p|, |IndexList|)
+    * The same constraints apply as for [=swizzle assignment=]:
+        * The indices in |IndexList| [=shader-creation error|must=] be all distinct from each other.
+        * Each index in |IndexList| [=shader-creation error|must=] be in [0,|N|-1].
```
Similarly: this constraint is already enforced by the "Swizzle View Expression" rules. Do we need to repeat it here? Should we just assert it instead?

jimblandy commented on wgsl/index.bs:
```diff
@@ -3790,17 +4060,34 @@ References and pointers are distinguished by how they are used:
 * A [=formal parameter=] can be of pointer type, but not of reference type.
+* A [=swizzle view=] can be formed by:
+    * applying a multi-letter [=swizzle=] name to a reference or pointer value whose [=store type=] is a vector.
+    * applying a multi-letter [=swizzle=] name to a swizzle view.
```
Does this belong here? This section is explaining what pointers and references are used for, and applying a multi-letter swizzle to a swizzle view doesn't consume or produce a pointer or reference, at least not explicitly.

jimblandy commented on wgsl/index.bs:
```diff
@@ -3790,17 +4060,34 @@ References and pointers are distinguished by how they are used:
+    * The [=right-hand side=] of the simple assignment statement [=shader-creation error|must=] evaluate to the store type of the left-hand side.
+* A [=swizzle assignment=] statement updates two or
+    more components of a vector in memory via a [=swizzle view=] that has
+    a [=access/read_write=] access mode.
```
Since this is the "reference and pointer use case" list, should this explicitly say how a pointer is involved?

dneto0 replied:
I added a sub-clause to describe the relationship again, and I reordered bullet points to bring the two swizzle-view clauses together.

<img width="766" height="185" alt="Screenshot 2026-07-03 at 3 51 38 PM" src="https://github.com/user-attachments/assets/8d6f2bd9-f9d5-4620-b004-e28228fdeafe" />

jimblandy commented on wgsl/index.bs:
Should these two "Pointer parameters" clauses say "memory views or swizzle views"?

dneto0 replied:
yes! Good catch.

jimblandy commented on wgsl/index.bs:
```diff
@@ -3650,6 +3674,244 @@ and vice versa,
+
+      // Indexing a swizzle view does not yield a swizzle view.
+      // Here v.xy is a swizzle value. There is no array-indexing expression on
+      // swizzle views, so the Swizzle Load Rule applies, yielding v.xy as a value
```
```suggestion
      // swizzle views, so the Swizzle View Load Rule applies, yielding v.xy as a value
```

jimblandy commented on wgsl/index.bs:
```diff
@@ -5877,97 +6178,181 @@ but a multi-letter swizzle expression always yields a value of [=vector=] type.
-    In this case the [=Load Rule=] applies, loading the vector component from memory and yielding that component as the result.
+* any place where a reference to a vector component can appear.
+    Sometimes the [=Load Rule=] will apply,
+    loading the vector component from memory and yielding that component as the result.
```
This reads strangely. I could be misunderstanding the intent, but it seems to me that, because this section is explaining vector component references, it's circular to say they can appear where a "reference to a vector component" can appear. I think what's intended here is more like:

```suggestion
* any place where a reference to the component's scalar value can appear.
    Since the [=Load Rule=] may apply, this includes any place where a scalar value can appear.
```

dneto0 replied:
That's better, thank you.
Applied.

jimblandy reviewed (approved):
Wow, fantastic!

dneto0 committed:
Apply low-level copyediting fixes from jimblandy

dneto0 committed:
SwizzleViewRead: give the types of the input and output

dneto0 committed:
Swizzle assignment: index range constraint is a note, not a constraint

It's already enforced by the swizzle view expression rules.

dneto0 committed:
remove "swizzle from swizzle" bullet point from "use cases for references and pointers"

dneto0 committed:
Logically link swizzle assignment to the original pointer or ref

"""
A swizzle assignment statement updates two or more components of a
vector in memory via a swizzle view that has a read_write access mode.

  A swizzle view is created from a pointer or reference, as above, or from
  another swizzle view. See § 8.15 Swizzle View Expression.
"""

dneto0 committed:
Add swizzle views to the pointer parameter cases for aliasing.

dneto0 committed:
Improve wording about where a component-reference can appear

dneto0 committed:
Swizzle views can only be read-write

- remove the access mode template parameter
- add the read-write constraint to the swizzle view type
  and the expressions that construct them.
- eliminate the access mode constraint on SwizzleViewWrite
  because it's now part of the swizzle view expression itself.

alan-baker commented on wgsl/index.bs:
```diff
@@ -3650,6 +3673,244 @@ and vice versa,
+<div class='example wgsl' heading='Reading and writing via swizzle views'>
+<span id="example-swizzle-reads-writes"></span>
+  <xmp highlight=wgsl>
+    fn swizzle_read_and_write() {
```
```suggestion
     requires pointer_composite_access;
     
     fn swizzle_read_and_write() {
```

dneto0 replied:
It's only really necessary for the second function.  I'll split it into two examples.

dneto0 replied:
actually, only the "same as ..." parts in the comments have to change.

The actual code itself doesn't change because p.xyz yields a swizzle view.

dneto0 replied:
I've incorporated this.   I split the sample into two samples, one on references, and the other with the pointer.

alan-baker commented on wgsl/index.bs:
```diff
@@ -3650,6 +3673,244 @@ and vice versa,
+      // of type vec2u. Then the indexed expression v.xy[0] results in a scalar value
+      // of type u32. That cannot be the left hand side of an assignment, since it is
+      // not a reference. Therefore, type checking fails.
+      v.xy[0] = 1; // Fails type checking.
```
This is different from other languages (e.g. GLSL).

dneto0 replied:
I talked about this at the fall 2025 F2F as a deliberate choice. See [slides](https://docs.google.com/presentation/d/1_gdPa8_XO8N43_2oaaiFVPZBkcgmUetmW9heVDtKh4E/edit?usp=sharing) slide 12, and again on slide 17 as a possible tweak.

12:
<img width="1041" height="552" alt="Screenshot 2026-07-06 at 5 25 52 PM" src="https://github.com/user-attachments/assets/a9920905-90d6-4cb4-8542-4f64aedad457" />

17:
<img width="533" height="454" alt="Screenshot 2026-07-06 at 5 26 11 PM" src="https://github.com/user-attachments/assets/67f31eea-bef2-40f1-b118-a3c986201b54" />

Also, issue 737 had a [comment](https://github.com/gpuweb/gpuweb/issues/737#issuecomment-1470282176) by John Stiles (then at Google) about the wobbliness of this feature in GLSL and Metal.

dneto0 replied:
The underlying shader languages do support it.

- GLSL: Glslang compiles it correctly to GLSL. (I read the GLSL spec and it wasn't 100% super clear that the indexing applied to lvalues that were swizzles..)
- HLSL /  DXC supports it
- MSL supports it.   I inspected the AIR assembly.


GLSL. Sample:
```
#version 450
shared uvec4 w;
void main() {
        w.yz[w.x] = 9;
}
```

HLSL. Sample:
```
groupshared uint4 w;

[numthreads(8,1,1)]
void main() {
  uint4 w;
  w.yz[w.x] = 1;
}
```

MSL.  (Compile with `xcrun metal -c a.metal -S -o -`)
```
#include <metal_stdlib>
using namespace metal;

[[max_total_threads_per_threadgroup(8)]]
kernel void v_1(device int* i) {
  uint4 v;
  v.yz[*i] = 1;
  *i = v.y;
}
```

alan-baker replied:
This is still marked as invalid. You updated the valid cases to include this, but forgot to delete this.

alan-baker commented on wgsl/index.bs:
```diff
@@ -2023,6 +2024,19 @@ Extension names are not [=identifiers=]: they do not [=resolves|resolve=] to [=d
+      a shorthand way of writing `v.x = 1; v.z = 2;`, while performing one read and one write.
+
+      This also works on pointers.  If `p` is a pointer to a vector of at least 3 elements, then
+      `p.xz = vec2(1,2);` is shorthand for `p.x = 1; p.z = 2;`.
```
This requires `pointer_composite_access`. Maybe better to not use it?

```suggestion
      `(*p).xz = vec2(1,2);` is shorthand for `(*p).x = 1; (*p).z = 2;`.
```

dneto0 replied:
Only change the shorthands.

dneto0 replied:
I'll rewrite as "if pointer_composite_access is also supported, then..."

alan-baker commented on wgsl/index.bs:
```diff
@@ -6778,6 +7226,113 @@ The <dfn noexport>indirection</dfn> operator converts a pointer to its correspon
+  </thead>
+  <tr algorithm="swizzle view from pointer expression">
+       <td>
+          [=language_extension/swizzle_assignment=] is supported
```
```suggestion
          [=language_extension/swizzle_assignment=] is supported<br>
          [=language_extension/pointer_composite_access=] is supported
```

dneto0 replied:
No, it's not dependent on `pointer_composite_access`

alan-baker replied:
That seems like mixing features to me.

dneto0 replied:
ok.  You convinced me. :-)

alan-baker commented on wgsl/index.bs:
```diff
@@ -6778,6 +7226,113 @@ The <dfn noexport>indirection</dfn> operator converts a pointer to its correspon
 
 </table>
 
+## Swizzle View Expression ## {#swizzle-view-expr}
```
Suggest moving this whole section just after composite decomposition.

dneto0 replied:
done

dneto0 committed:
Don't depend on pointer_composite_access in description of swizzle expressions

Instead of saying p.yz = ... is the "same as" p.y = ... and p.z = ...
say (*p).y = ... and (*p).z = ...

dneto0 committed:
Move "Swizzle View Expression" to just after "Composite Value Decomposition Expressions"

jimblandy commented on wgsl/index.bs:
```diff
@@ -3791,16 +4060,33 @@ References and pointers are distinguished by how they are used:
+    with [=access/read_write=] access mode and whose [=store type=] is a vector.
+* A [=swizzle assignment=] statement updates two or
+    more components of a vector in memory via a [=swizzle view=]
+    a [=access/read_write=] access mode.
```
```suggestion
    more components of a vector in memory via a [=swizzle view=].
```

jimblandy reviewed (approved):
Looks good to me. One minor fixup:

dneto0 committed:
Allow array-indexing of swizzle: v.yz[i]

dneto0 committed:
Remove dangling sentence fragment (thanks jim)

dneto0 committed:
Swizzle from pointer requires pointer_composite_access

dneto0 committed:
Fix typos in swizzle array accessing expression description

dneto0 committed:
Remove old example of invalid indexed swizzle

dneto0 committed:
Add spacing for legibility

dneto0 committed:
Tweak wording of swizzle view expression section

Provide additional scaffolding words:
- Change "swizzle Z" to "swizzle name Z" for clarity
- Change "I_in" to "the sequence I_n"

Fix constraint on K:
  Before:  K in {2,3,4}
  Fixed:   K_Z in {2,3,4}

dneto0 committed:
Spell out index sequences in swizzle view expr constraints

jimblandy reviewed (approved):
All new changes look great.

dneto0 commented:
I rebased against main, resolving conflicts due to other language features coming in since the last rebase.

dneto0 commented:
The build failure seems triggered by a new Bikeshed warning for markdown-style blockquotes.
I'll post another PR to convert those, then rebase on top of that.

dneto0 commented:
Posted #6878 to fix the blockquote issue
